### PR TITLE
CLOUDP-283049: fix data federation endpoint reconciliation

### PR DESCRIPTION
**Pull request by s-urbaniak**
This is an alternative to #1928 and introduces a state machine handling.

It also prevents accidental deletion of data federation private endpoints that are not tracked by AKO.

### All Submissions:

* [ ] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: internal/translation/datafederation/conversion_endpoints.go [hunk 1]
```diff
@@ -58,3 +65,39 @@ func endpointToAtlas(ep *DatafederationPrivateEndpointEntry) *admin.PrivateNetwo
...
+}
+
+func MapDatafederationPrivateEndpoints(projectID string, df *akov2.AtlasDataFederation, atlasEndpoints []*DatafederationPrivateEndpointEntry) (map[string]*DataFederationPrivateEndpoint, error) {
+	var lastApplied akov2.AtlasDataFederation
+	if ann, ok := df.GetAnnotations()[customresource.AnnotationLastAppliedConfiguration]; ok {
+		err := json.Unmarshal([]byte(ann), &lastApplied.Spec)
+		if err != nil {
+			return nil, fmt.Errorf("error reading data federation from last applied annotation: %w", err)
+		}
+	}
```
Comment (josvazg) [inline internal/translation/datafederation/conversion_endpoints.go:80]: nit

Being purist I would do this outside of the translation layer. Basically the translation layer should be Kubernetes unaware. We could pass this as an argument.

Not blockling the PR due to this, but design wise I think is better to not tangle the translation with the concept of annotations in kube objects.

File: internal/translation/datafederation/conversion_endpoints.go [hunk 2]
```diff
@@ -58,3 +65,39 @@ func endpointToAtlas(ep *DatafederationPrivateEndpointEntry) *admin.PrivateNetwo
...
+
+	result := make(map[string]*DataFederationPrivateEndpoint)
+	addEndpoint := func(endpointID string) {
+		if _, ok := result[endpointID]; !ok {
+			result[endpointID] = &DataFederationPrivateEndpoint{}
+		}
+	}
```
Comment (josvazg) [inline internal/translation/datafederation/conversion_endpoints.go:87]: Why was a closure preferred to a simple external private function such as this?
```go
func(result map[string]*DataFederationPrivateEndpoint, endpointID string) {
	if _, ok := result[endpointID]; !ok {
		result[endpointID] = &DataFederationPrivateEndpoint{}
	}
}
```


Closures are required when the closure needs to travel to another context, bringing its context with it. This is not the case.

Comment (s-urbaniak) [reply internal/translation/datafederation/conversion_endpoints.go:87]: This closes over `result`

**Review by igor-karpukhin — APPROVED** [review]

File: internal/translation/datafederation/conversion_endpoints.go (hunk 1 above)
Comment (s-urbaniak) [reply internal/translation/datafederation/conversion_endpoints.go:80]: I actually don't agree in this case.

This is definitely part of the translation layer, it introspects public types and its internal state, in this case annotation, and generates internal types to be consumed by the controller, in this case a map.

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 3]
```diff
@@ -116,8 +116,8 @@ func (r *AtlasDataFederationReconciler) Reconcile(context context.Context, req c
 		return result.ReconcileResult(), nil
 	}
 
-	if result = r.ensurePrivateEndpoints(ctx, project, dataFederation, endpointService); !result.IsOk() {
-		ctx.SetConditionFromResult(api.DataFederationPEReadyType, result)
+	if result = r.ensurePrivateEndpoints(ctx, endpointService, project, dataFederation); !result.IsOk() {
```
Comment (josvazg) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:119]: why the order changes?

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 4]
```diff
@@ -116,8 +116,8 @@ func (r *AtlasDataFederationReconciler) Reconcile(context context.Context, req c
...
+		ctx.SetConditionFromResult(api.DataFederationReadyType, result)
```
Comment (josvazg) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:120]: This was another bug?

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 5]
```diff
@@ -145,7 +145,7 @@ func (r *AtlasDataFederationReconciler) Reconcile(context context.Context, req c
 		return r.handleDelete(ctx, log, dataFederation, project, dataFederationService).ReconcileResult(), nil
 	}
 
-	err = customresource.ApplyLastConfigApplied(context, project, r.Client)
+	err = customresource.ApplyLastConfigApplied(context, dataFederation, r.Client)
```
Comment (josvazg) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:148]: Seems like another bug fixed here right?

File: internal/translation/datafederation/conversion_endpoints.go (hunk 1 above)
Comment (s-urbaniak) [reply internal/translation/datafederation/conversion_endpoints.go:80]: > the translation layer should be Kubernetes unaware. 

I agree but this is not the case here. We are introspecting the structure of the public type. The annotation is a simple map, we encode previous state in here using JSON, there is nothing Kubernetes specific in here.

File: pkg/controller/atlasdatafederation/private_endpoint.go [hunk 6]
```diff
@@ -1,64 +1,103 @@
...
 	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/datafederation"
-
-	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/set"
 	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
 	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
 	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
 )
 
-func (r *AtlasDataFederationReconciler) ensurePrivateEndpoints(ctx *workflow.Context, project *akov2.AtlasProject, dataFederation *akov2.AtlasDataFederation, service datafederation.DatafederationPrivateEndpointService) workflow.Result {
+func (r *AtlasDataFederationReconciler) ensurePrivateEndpoints(ctx *workflow.Context, service datafederation.DatafederationPrivateEndpointService, project *akov2.AtlasProject, dataFederation *akov2.AtlasDataFederation) workflow.Result {
```
Comment (josvazg) [inline pkg/controller/atlasdatafederation/private_endpoint.go:12]: I would not change order of params, it is an internal API and does not matter much, but still there is no reason for the change AFAIK

File: pkg/controller/atlasdatafederation/datafederation_controller.go (hunk 5 above)
Comment (s-urbaniak) [reply pkg/controller/atlasdatafederation/datafederation_controller.go:148]: yes, it's part of the ticket being solved.

File: pkg/controller/atlasdatafederation/private_endpoint.go (hunk 6 above)
Comment (s-urbaniak) [reply pkg/controller/atlasdatafederation/private_endpoint.go:12]: see https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1929#discussion_r1846084044

Comment (s-urbaniak) [reply pkg/controller/atlasdatafederation/private_endpoint.go:12]: I actually suggest to change the variable ordering, it makes the signature clearer.

File: pkg/controller/atlasdatafederation/private_endpoint.go [hunk 7]
```diff
@@ -1,64 +1,103 @@
...
+	ctx.UnsetCondition(api.DataFederationPEReadyType)
+	return workflow.OK()
+}
+
+type PrivateEndpointReconciler struct {
+	service  datafederation.DatafederationPrivateEndpointService
+	endpoint *datafederation.DataFederationPrivateEndpoint
+}
+
+func (r *PrivateEndpointReconciler) Reconcile(ctx *workflow.Context) error {
```
Comment (josvazg) [inline pkg/controller/atlasdatafederation/private_endpoint.go:60]: It would be good to point to the state machine diagram doc somehow

Comment (s-urbaniak) [reply pkg/controller/atlasdatafederation/private_endpoint.go:60]: you mean the dot plot? I think this is a very simple state machine, I suggest to address this as a follow-up if you believe it is worth doing.

File: internal/translation/datafederation/conversion_endpoints.go (hunk 1 above)
Comment (josvazg) [reply internal/translation/datafederation/conversion_endpoints.go:80]: Actually, you are right, we are not using any Kubernetes interfaces or APIs here. Annotations are just maps, so yes, I guess is fine.

**Pull request merged**